README.rst
.. contents::


Documentation
=============

.. image:: http://img.shields.io/pypi/v/collective.catalogcleanup.svg
   :target: https://pypi.python.org/pypi/collective.catalogcleanup


Usage and goal
--------------

Add ``collective.catalogcleanup`` to the eggs in your buildout.
This makes a browser view available on the Plone Site root: ``@@collective-catalogcleanup``.

This goes through the ``portal_catalog`` and removes all catalog brains for which a ``getObject`` call does not work.
In other words, it removes brains that no longer belong to an actual object in the site.

The goal is to get rid of outdated brains that could otherwise cause problems, for example during an upgrade to a new major Plone version.

``@@collective-catalogcleanup`` by default does a dry run, so it only reports problems.
Call it with ``@@collective-catalogcleanup?dry_run=false`` to perform the actual cleanup.


Details
-------

So what does the catalog cleanup do?

- It removes stuff!  You *must* make a backup first!

- It handles the ``portal_catalog``.
  Version 1.x also handled ``uid_catalog`` and ``reference_catalog``.

- For each catalog it reports the number of catalog brains it contains.

- It removes brains that have a UID of ``None``.

- It removes brains of which the object is broken.  This can happen
  when the object belongs to a package that is no longer available in
  the Plone Site.

- It removes brains of which the object cannot be found.

- It looks for non unique uids.
  We accept one object and we give the other objects a new UID.

- A simple report will be printed in the browser.
  It may look like this::

    Brains in portal_catalog: 20148
    portal_catalog: removed 25 brains without UID.
    portal_catalog: removed 100 brains with status broken.
    portal_catalog: removed 5 brains with status notfound.
    portal_catalog: 249 non unique uids found.
    portal_catalog: 249 items given new unique uids.

- The instance log may contain more info, about individual items.


Alternatives
------------

- A clear and rebuild of the ``portal_catalog`` should have partially the
  same effect, but it will likely take a lot longer and it will not
  solve some of the problems mentioned above.  But this is definitely
  the most logical thing to try before giving
  ``collective.catalogcleanup`` a go.


Compatibility
-------------

Version 2.x works on Plone 5.2 and 6.0, on Python 3 only.
For earlier Plone and Python versions, use version 1.x.


Authors
-------

Maurits van Rees


CHANGES.rst
Changelog
=========


2.0.1 (unreleased)
------------------

- Nothing changed yet.


2.0.0 (2023-11-30)
------------------

- No longer accept non-unique inherited UIDs.
  This was needed for Discussion Items, but not anymore.
  [maurits]

- Fixed setting new uuids for Dexterity content and Discussion Items.
  This part of version 1.x seems to only have worked for Archetypes.
  [maurits]

- Removed support for checking the ``reference_catalog`` and ``uid_catalog``.
  These are Archetypes-only.
  [maurits]

- Drop support for Python 2, Plone 5.1 and earlier.
  Only Plone 5.2 and 6 are supported, on Python 3.
  [maurits]


1.11.2 (2021-08-24)
-------------------

- Do not complain about paths being different when the brain path is relative.
  Happens in old uid_catalog and reference_catalog.
  [maurits]


1.11.1 (2021-08-13)
-------------------

- Again fixed error sorting UIDs when one of them is None.  [maurits]


1.11.0 (2021-07-29)
-------------------

- Report and remove brains that have a wrong path.
  These are brains where an object is found at the given path,
  but the actual path of this object is different.
  This can be caused by acquisition:
  /Plone/folder/folder has gotten in the catalog,
  but really only /Plone/folder exists.
  [maurits]

- Fixed error sorting UIDs when one of them is None.  [maurits]


1.10.0 (2020-02-08)
-------------------

- To get all brains, use ``catalog.getAllBrains`` when available.
  This is available in ``Products.ZCatalog`` 2.13.30+ or 4.1+.
  We fall back to ``unrestrictedSearchResults`` and otherwise just call the catalog.
  Fixes `issue 16 <https://github.com/collective/collective.catalogcleanup/issues/16>`_.
  [maurits]

- `Bug 22 <https://github.com/collective/collective.catalogcleanup/issues/22>`_:
  Fix startup error when Archetypes is missing.
  Fix other Plone 5.2 and Python 3 errors.
  Should be compatible with Plone 4.3, 5.1 and 5.2 now.  [maurits]


1.9.0 (2018-09-25)
------------------

- Catch TypeError when getting object for brain.
  Can happen when an object that used to be referenceable is no longer referenceable.
  Fixes `issue #19 <https://github.com/collective/collective.catalogcleanup/issues/19>`_.
  [maurits]

- Disable CSRF protection.
  Fixes `issue #17 <https://github.com/collective/collective.catalogcleanup/issues/17>`_.
  [maurits]

- Abort any transaction changes in dry run mode.
  There should not be any changes here anyway, but this makes sure.
  [maurits]


1.8.0 (2018-04-30)
------------------

- No longer test on Plone 4.1 and 4.2 and on Python 2.6.  [maurits]

- Catch ``KeyError`` and ``AttributeError`` for ``getPath`` in more cases.
  Fixes `issue #14 <https://github.com/collective/collective.catalogcleanup/issues/14>`_.
  [maurits]


1.7.2 (2017-09-18)
------------------

- Added traceback info to help in case of problems.  [maurits]


1.7.1 (2017-03-07)
------------------

- Tested for compatibility on Plone 4.0 through 5.1.  [hvelarde]

- Ignore non existing catalogs.  Plone 5 does not always have
  a ``uid_catalog`` or ``reference_catalog``.
  Fixes `issue #5 <https://github.com/collective/collective.catalogcleanup/issues/5>`_.
  [maurits]


1.7 (2017-03-03)
----------------

- Don't look for non unique ids in the ``reference_catalog``.
  It looks like it is normal there.  At least, on one Plone 4.3 site
  the code keeps creating several new uids every time I run it.
  [maurits]

- Don't complain about brains in ``reference_catalog`` where ``getObject`` returns None.
  This happens for content without apparent problems.  [maurits]


1.6 (2016-08-23)
----------------

- Do not complain about brains in uid_catalog that are references.
  When their path points to ``...at_references/<uid of brain>`` then
  this is normal.  I started wondering about a site that had more than
  20 thousand problems reported this way.  [maurits]


1.5 (2015-07-31)
----------------

- Remove all items that have the ``portal_factory`` folder in their
  path.
  [maurits]


1.4 (2014-05-12)
----------------

- Catch KeyErrors when getting the path of a brain.
  [maurits]


1.3 (2013-09-02)
----------------

- Give less confusing message for comments that inherit the UID of
  their parent.  It sounded too much like an error.
  [maurits]


1.2 (2012-06-04)
----------------

- Improved the cleanup of non unique uids.
  [maurits]


1.1 (2012-05-14)
----------------

- When doing an reindexObject, only reindex the UID.
  [maurits]


1.0 (2012-04-27)
----------------

- Initial release
  [maurits]


collective/catalogcleanup/tests/test_functional.py
from collective.catalogcleanup.testing import (  # noqa: E501
    CATALOG_CLEANUP_FUNCTIONAL_TESTING,
)
from collective.catalogcleanup.testing import cleanup
from collective.catalogcleanup.testing import make_test_doc
from collective.noindexing import patches
from plone.app.testing import setRoles
from plone.app.testing import TEST_USER_ID
from Products.CMFCore.utils import getToolByName

import transaction
import unittest


class TestCatalogCleanup(unittest.TestCase):
    layer = CATALOG_CLEANUP_FUNCTIONAL_TESTING

    def _makeOne(self):
        return make_test_doc(self.layer["portal"])

    def _delete_object_only(self, doc):
        # Delete object without removing it from the catalog.
        portal = self.layer["portal"]
        patches.apply()
        portal._delObject(doc.getId())
        patches.unapply()
        transaction.commit()

    def testNormalDeletedDocument(self):
        # No tricks here, just testing some assumptions.
        portal = self.layer["portal"]
        setRoles(portal, TEST_USER_ID, ("Manager",))
        catalog = getToolByName(portal, "portal_catalog")
        base_count = len(catalog.searchResults({}))
        doc = self._makeOne()
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        portal._delObject(doc.getId())
        transaction.commit()
        self.assertEqual(len(catalog.searchResults({})), base_count)
        cleanup(portal)
        self.assertEqual(len(catalog.searchResults({})), base_count)

    def testDeletedDocumentWithDryRun(self):
        portal = self.layer["portal"]
        setRoles(portal, TEST_USER_ID, ("Manager",))
        catalog = getToolByName(portal, "portal_catalog")
        base_count = len(catalog.searchResults({}))
        doc = self._makeOne()
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        # This call makes sure the item remains in the catalog after
        # it is removed:
        self._delete_object_only(doc)
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        # By default dry_run is selected, so nothing is changed.
        cleanup(portal)
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        # None of these variants should have any lasting effect.
        cleanup(portal, dry_run=True)
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        cleanup(portal, dry_run=0)
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        cleanup(portal, dry_run=None)
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)

    def testDeletedDocumentForReal(self):
        portal = self.layer["portal"]
        setRoles(portal, TEST_USER_ID, ("Manager",))
        catalog = getToolByName(portal, "portal_catalog")
        base_count = len(catalog.searchResults({}))
        doc = self._makeOne()
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        # This call makes sure the item remains in the catalog after
        # it is removed:
        self._delete_object_only(doc)
        self.assertEqual(len(catalog.searchResults({})), base_count + 1)
        cleanup(portal, dry_run="false")
        self.assertEqual(len(catalog.searchResults({})), base_count)


collective/catalogcleanup/testing.py
from plone.app.testing import FunctionalTesting
from plone.app.testing import PLONE_FIXTURE
from plone.app.testing import PloneSandboxLayer
from zope.component import getMultiAdapter

import random
import string
import transaction


class CatalogCleanupLayer(PloneSandboxLayer):
    defaultBases = (PLONE_FIXTURE,)

    def setUpZope(self, app, configurationContext):
        import collective.catalogcleanup

        self.loadZCML(package=collective.catalogcleanup)


CATALOG_CLEANUP_FIXTURE = CatalogCleanupLayer()
CATALOG_CLEANUP_FUNCTIONAL_TESTING = FunctionalTesting(
    bases=(CATALOG_CLEANUP_FIXTURE,), name="CatalogCleanup:Integration"
)

# A few helper functions.


def make_test_doc(portal):
    # new_id = portal.generateUniqueId('Document')
    # generateUniqueId no longer exists in all Plone versions.
    # Get a sufficiently unique id.
    new_id = "doc_{}_{}".format(
        "".join(random.sample(string.ascii_letters, 10)),
        "".join(random.sample(string.digits, 4)),
    )
    portal.invokeFactory("Document", new_id)
    doc = portal[new_id]
    doc.reindexObject()  # Might have already happened, but let's be sure.
    transaction.commit()
    return doc


def cleanup(portal, **kwargs):
    view = getMultiAdapter((portal, portal.REQUEST), name="collective-catalogcleanup")
    return view(**kwargs)


collective/catalogcleanup/browser.py
from Acquisition import aq_inner
from itertools import groupby
from OFS.Uninstalled import BrokenClass
from plone.protect.interfaces import IDisableCSRFProtection
from plone.uuid.interfaces import ATTRIBUTE_NAME
from plone.uuid.interfaces import IUUID
from plone.uuid.interfaces import IUUIDGenerator
from Products.CMFCore.utils import getToolByName
from Products.Five import BrowserView
from zExceptions import NotFound
from ZODB.POSException import ConflictError
from zope.component import queryUtility
from zope.interface import alsoProvides

import logging
import transaction


logger = logging.getLogger("collective.catalogcleanup")


def safe_path(item):
    try:
        return item.getPath()
    except (KeyError, AttributeError):
        return "notfound"


def path_len(item):
    try:
        return len(item.getPath())
    except (KeyError, AttributeError):
        # For our use case (sort by path length and keep the item
        # with the shortest path), it is best to return a high number
        # in case of problems.
        return 9999


def get_all_brains(catalog):
    # Return a full list, not a generator,
    # because we may delete brains from the catalog while iterating over them.
    return list(catalog.getAllBrains())


def uid_getter(brain):
    """Get UID from brain.

    operator.attrgetter('UID') should be fine,
    but this fails for sorting when there one of the UIDs is None:

    '<' not supported between instances of 'NoneType' and 'str'

    """
    return getattr(brain, "UID", "") or ""


class Cleanup(BrowserView):
    def __call__(self, dry_run=None):
        # Avoid csrf protection errors, as we have no form.
        alsoProvides(self.request, IDisableCSRFProtection)
        self.messages = []
        self.msg("Starting catalog cleanup.")
        # Determine whether this is a dry run or not.  We are very
        # explicit and only accept the boolean value False and the
        # string 'false' (in lower, upper or mixed case).  All other
        # values are considered True.
        if dry_run is None:
            dry_run = self.request.get("dry_run")
        if isinstance(dry_run, str):
            dry_run = dry_run.lower()
            if dry_run == "false":
                dry_run = False
        if dry_run is False:
            self.dry_run = False
            self.newline()
            self.msg("NO dry_run SELECTED. CHANGES ARE PERMANENT.")
            self.newline()
        else:
            self.dry_run = True
            self.newline()
            self.msg(
                "dry_run SELECTED, SO ONLY REPORTING. To make changes "
                'permanent, add "?dry_run=false" to the URL.'
            )
            self.newline()
        context = aq_inner(self.context)
        catalog_ids = ["portal_catalog"]
        for catalog_id in catalog_ids:
            problems = 0
            self.newline()
            if getToolByName(context, catalog_id, None) is None:
                self.msg(f"Ignored non existing catalog {catalog_id}.")
                continue
            self.msg(f"Handling catalog {catalog_id}.")
            problems += self.report(catalog_id)
            problems += self.remove_without_uids(catalog_id)
            problems += self.remove_without_object(catalog_id)
            problems += self.non_unique_uids(catalog_id)
            self.msg(f"{catalog_id}: total problems: {problems:d}")

        self.newline()
        self.msg("Done with catalog cleanup.")
        if self.dry_run:
            # We should not have made any changes, but let's back out any
            # inadvertent changes anyway.
            transaction.abort()
            self.msg("Dry run selected: aborted any transaction changes.")

        return "\n".join(self.messages)

    def msg(self, msg, **kwargs):
        level = kwargs.get("level", logging.INFO)
        logger.log(level, msg)
        self.messages.append(msg)

    def newline(self):
        self.messages.append("")

    def report(self, catalog_id):
        """Report about this catalog."""
        __traceback_info__ = catalog_id
        context = aq_inner(self.context)
        catalog = getToolByName(context, catalog_id)
        # This uses a special method to get the length:
        size = len(catalog)
        self.msg(f"Brains in {catalog_id}: {size:d}")
        # Getting all brains from the catalog may give a different
        # result for various reasons: multilingual language filter,
        # publication date in future, catalog not returning any results
        # when called without a query, even with unrestrictedSearchResults.
        # So check that we can get all brains, as we will use this in all checks.
        alternative_size = len(get_all_brains(catalog))
        if alternative_size != size:
            self.msg(
                "Brains in {} using getAllBrains is different: {:d}".format(
                    catalog_id, alternative_size
                ),
                level=logging.WARN,
            )
            return 1
        return 0

    def remove_without_uids(self, catalog_id):
        """Remove all brains without UID."""
        __traceback_info__ = catalog_id
        context = aq_inner(self.context)
        catalog = getToolByName(context, catalog_id)
        uncatalog = 0
        # We need to get the complete list instead of a lazy
        # mapping, otherwise iterating misses half of the brains
        # and we would need to try again.
        brains = list(catalog.unrestrictedSearchResults(UID=None))
        for brain in brains:
            if not self.dry_run:
                try:
                    path = brain.getPath()
                except (KeyError, AttributeError):
                    continue
                catalog.uncatalog_object(path)
            uncatalog += 1
        self.msg(f"{catalog_id}: removed {uncatalog:d} brains without UID.")
        return uncatalog

    def remove_without_object(self, catalog_id):
        """Remove all brains without object."""
        __traceback_info__ = catalog_id
        context = aq_inner(self.context)
        catalog = getToolByName(context, catalog_id)
        status = {}
        brains = get_all_brains(catalog)
        for brain in brains:
            obj = self.get_object_or_status(brain)
            if not isinstance(obj, str):
                continue
            if not self.dry_run:
                try:
                    path = brain.getPath()
                except (KeyError, AttributeError):
                    continue
                catalog.uncatalog_object(path)
            # We have an error.
            count = status.get(obj, 0)
            status[obj] = count + 1

        for error, value in status.items():
            self.msg(
                "{}: removed {:d} brains with status {}.".format(
                    catalog_id, value, error
                )
            )
        if not status:
            self.msg(f"{catalog_id}: removed no brains in object check.")
        return sum(status.values())

    def non_unique_uids(self, catalog_id):
        """Report and fix non unique uids.

        A UID that is not unique is wrong.  And the UUIDIndex
        migration that is done when migrating from Plone 3 to Plone 4
        breaks in that case.
        """
        __traceback_info__ = catalog_id
        context = aq_inner(self.context)
        catalog = getToolByName(context, catalog_id)
        if "UID" not in catalog.indexes():
            self.msg(f"{catalog_id}: no UID index.")
            return
        non_unique = 0
        changed = 0
        obj_errors = 0
        # Get a uuid generator and try it out.
        uuid_generator = queryUtility(IUUIDGenerator)
        try_uuid = uuid_generator() if uuid_generator is not None else None
        if not try_uuid:
            self.msg("Could not get a working uuid generator utility.")
            return
        brains = get_all_brains(catalog)
        brains = sorted(brains, key=uid_getter)
        for uid, group in groupby(brains, uid_getter):
            items = list(group)
            if len(items) == 1:
                continue
            non_unique += 1
            logger.info("%s: uid %s: %d items.", catalog_id, uid, len(items))
            # Sort by length of path.
            items = sorted(items, key=path_len)
            logger.info(
                "%s: uid %s is kept for %s", catalog_id, uid, safe_path(items[0])
            )
            for item in items[1:]:
                obj = self.get_object_or_status(item)
                if isinstance(obj, str):
                    # This is an error. This should fix itself when no
                    # dry_run has been selected.
                    obj_errors += 1
                    continue
                if obj is None:
                    # No error, but no object either.
                    continue
                old_uid = IUUID(obj, None)
                if old_uid is None:
                    # Comments used to inherit the UID of their parent, at
                    # least in Plone 3.  This is no longer the case in Plone 5.2/6.0.
                    # This may have already changed since 4.1 with the introduction of
                    # plone.app.discussion.  We used to accept this and do a reindex
                    # of the UID, as we may have given the parent a fresh UID a moment
                    # ago.  But now we warn, and create a new uuid.
                    logger.info(
                        "%s: %s has no uid of its own. Will give it a fresh one.",
                        catalog_id,
                        safe_path(item),
                    )
                # We need a change.
                changed += 1
                if not self.dry_run:
                    uuid = uuid_generator()
                    setattr(obj, ATTRIBUTE_NAME, uuid)
                    obj.reindexObject(idxs=["UID"])
                    logger.info(
                        "{}: new uid {} for {} (was {}).".format(
                            catalog_id, obj.UID(), safe_path(item), old_uid
                        )
                    )

        if obj_errors:
            self.msg(f"{catalog_id}: problem getting {obj_errors:d} objects.")
        self.msg(f"{catalog_id}: {non_unique:d} non unique uids found.")
        if self.dry_run:
            self.msg(f"{catalog_id}: {changed:d} items need new unique uids.")
        else:
            self.msg(f"{catalog_id}: {changed:d} items given new unique uids.")
        return obj_errors + changed

    def get_object_or_status(self, brain, getter="getObject"):
        __traceback_info__ = [brain, getter]
        try:
            brain_id = brain.getPath()
        except (AttributeError, KeyError):
            return "notfound"
        else:
            if "portal_factory" in brain_id.split("/"):
                return "factory"
        __traceback_info__.append(brain_id)
        try:
            # Usually: brain.getObject()
            obj = getattr(brain, getter)()
        except (ConflictError, KeyboardInterrupt):
            raise
        except (NotFound, AttributeError, KeyError):
            return "notfound"
        except TypeError:
            logger.warning(
                "TypeError, returning notfound for brain at %s.", brain_id, exc_info=1
            )
            return "notfound"
        except Exception:
            logger.exception("Cannot handle brain at %s.", brain_id)
            raise
        if obj is None:
            return "none"
        if isinstance(obj, BrokenClass):
            logger.warning("Broken %s: %s", brain.portal_type, brain_id)
            return "broken"
        if brain_id.startswith("/"):
            actual_path = "/".join(obj.getPhysicalPath())
            if brain_id != actual_path:
                # Likely acquisition:
                # /Plone/folder/folder has gotten in the catalog,
                # but really only /Plone/folder exists.
                logger.warning(
                    "Wrong path: getting %s leads to %s", brain_id, actual_path
                )
                return "wrong_path"
        return obj
